### PR TITLE
WT-6280 Fail eviction if out of order handling races with checkpoint

### PR DESCRIPTION
**Pull request by quchenhao**
The solution is to fail eviction when we detect either mixed mode or out of order updates when checkpoint is running.

### CODE REVIEW

**Comment by quchenhao** [comment]
There is a new test failure I need to take a look.

**Comment by quchenhao** [comment]
The test failure is caused by WT-7771. I think my change itself is OK.

**Comment by quchenhao** [comment]
@luke-pearson @bvpvamsikrishna @agorrod Can you have a look at the change?

File: src/evict/evict_page.c
Comment (luke-pearson) [inline src/evict/evict_page.c:666]: I can't remember but we don't need to lock to access this?

File: src/evict/evict_page.c
Comment (luke-pearson) [inline src/evict/evict_page.c:684]: Isn't there a risk of gcc optimizing this out if we don't use a barrier?

File: src/evict/evict_page.c
Comment (quchenhao) [reply src/evict/evict_page.c:666]: @hbokhari Can you explain why there is no barrier?

File: src/evict/evict_page.c
Comment (quchenhao) [reply src/evict/evict_page.c:684]: This is in the same thread. The first read must happen before this read.

Comment (quchenhao) [reply src/evict/evict_page.c:684]: I don't think gcc can do this.

File: src/history/hs_rec.c
Comment (luke-pearson) [inline src/history/hs_rec.c:207]: Add statistic for this eviction failure case.

File: src/evict/evict_page.c
Comment (quchenhao) [reply src/evict/evict_page.c:666]: Generally I think this is fine. There is only one place we set this flag to true, which is at the start of the checkpoint.

Comment (quchenhao) [reply src/evict/evict_page.c:666]: The other threads are just reading the value.

File: src/evict/evict_page.c
Comment (luke-pearson) [inline src/evict/evict_page.c:698]: I think setting this flag here is correct but I want to make sure. The error case happens when eviction evicts pages from a btree and checkpoint is yet to get to the history store, and must have already processed this btree? Can the btree that eviction evicts be ahead of the checkpoint and if so would that be an error case or fine?

Comment (quchenhao) [reply src/evict/evict_page.c:698]: Currently, we can't evict ahead of checkpoint. If we can, that is still run, because they use different snapshot. Eviction may see the updates checkpoint cannot see.

File: src/evict/evict_page.c
Comment (luke-pearson) [reply src/evict/evict_page.c:666]: Yeah but the order of these reads appears to be important, I think gcc can re-order non volatile reads but as this is a volatile it should be okay?

File: src/evict/evict_page.c
Comment (luke-pearson) [reply src/evict/evict_page.c:698]: Cool thanks.

File: src/evict/evict_page.c
Comment (quchenhao) [reply src/evict/evict_page.c:666]: I think the order only matters for the write thread and read thread? It doesn't matter between reads. Within this thread, the read cannot be reordered.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:207]: Added in 7cacea7

File: src/evict/evict_page.c
Comment (hbokhari) [reply src/evict/evict_page.c:666]: I believe the variable `conn->txn_global.checkpoint_running` is only being set in the checkpoint thread and only get tread in other threads. There is no need for lock in that scenario.

**Comment by agorrod** [comment]
This is in my radar. I'll get to it early next week

File: src/evict/evict_page.c
Comment (agorrod) [inline src/evict/evict_page.c:684]: I don't think we need a lock, but Luke's comment seems on point. The comment here says "A new checkpoint might have started between previous check and acquiring snapshot", which indicates that this code really wants to read the global variable again, not use a stashed local copy. I believe it won't need any explicit controls, since bumping the snapshot should already have a barrier in it (though that's worth checking).

File: src/evict/evict_page.c
Comment (agorrod) [inline src/evict/evict_page.c:698]: I think you are repurposing that flag and I don't think that's the right approach. It should be OK for an eviction in a tree other than the one being checkpointed right now to evict content in front of the checkpoint snapshot. We did a lot of work to make that possible. If we disallow that all content will need to be kept in cache for the duration of the checkpoint.
The content being evicted in the tree currently being checkpointed is only evicted by the checkpoint thread - so the correct flag should already be set.

File: src/history/hs_rec.c
Comment (agorrod) [inline src/history/hs_rec.c:443]: Thanks for adding stats!

File: src/history/hs_rec.c
Comment (agorrod) [inline src/history/hs_rec.c:791]: I'm a bit apprehensive that this change will introduce failure paths in reconciliation we haven't tested much before. This change deserves extensive format and MongoDB testing before merge.

File: src/evict/evict_page.c
Comment (agorrod) [reply src/evict/evict_page.c:684]: What semantic do we need here exactly? Do we need to know if a checkpoint was running when the snapshot was allocated for this reconciliation? Would it be OK to know if there was a checkpoint running immediately before or immediately after?

If a checkpoint started and finished between that window, could this reconciliation possibly rewrite content in the history store?

**Review by agorrod — COMMENTED** [review]
I've got some questions and concerns about this change - I've tried to make them clear in the comments - let me know if they require more context

File: src/evict/evict_page.c
Comment (quchenhao) [reply src/evict/evict_page.c:684]: If the checkpoint finished in the middle, the history store should have been already checkpointed correctly. So that case should be OK. Here we want to know if there is a checkpoint on going before we get the eviction snapshot. If there is one, we give up getting the snapshot. Then we check again after we have got the snapshot, if there is one started concurrently, we release the snapshot to go for the conservative approach. If we don't detect a checkpoint, it either means there is no one started concurrently, which is OK, or one have started and finished already, which is also OK. So it is safe to use the eviction snapshot.

File: src/evict/evict_page.c
Comment (quchenhao) [reply src/evict/evict_page.c:698]: I'm not repurposing the flag. It is a new flag I added in this change.

Comment (quchenhao) [reply src/evict/evict_page.c:698]: bq. It should be OK for an eviction in a tree other than the one being checkpointed right now to evict content in front of the checkpoint snapshot. We did a lot of work to make that possible. If we disallow that all content will need to be kept in cache for the duration of the checkpoint.

I'm not holding up all the content to be evicted. The change only fails eviction if there is out of order timestamp updates in that page. If there is none, it can be evicted.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:791]: I have ran mongodb patch build and sys-perf patch build and the links are on the ticket.

File: src/evict/evict_page.c
Comment (quchenhao) [reply src/evict/evict_page.c:698]: I believe it is wrong to evict pages with out of order timestamp updates ahead of checkpoint as rollback to stable in this case doesn't see the original timestamp of the updates and may restore the wrong value.

File: src/evict/evict_page.c
Comment (quchenhao) [reply src/evict/evict_page.c:684]:
```
/* Walk the array of concurrent transactions. */
    WT_ORDERED_READ(session_cnt, conn->session_cnt);
    WT_STAT_CONN_INCR(session, txn_walk_sessions);
    for (i = 0, s = txn_global->txn_shared_list; i < session_cnt; i++, s++) {
        WT_STAT_CONN_INCR(session, txn_sessions_walked);
        /*
         * Build our snapshot of any concurrent transaction IDs.
         *
         * Ignore:
         *  - Our own ID: we always read our own updates.
         *  - The ID if it is older than the oldest ID we saw. This
         *    can happen if we race with a thread that is allocating
         *    an ID -- the ID will not be used because the thread will
         *    keep spinning until it gets a valid one.
         *  - The ID if it is higher than the current ID we saw. This
         *    can happen if the transaction is already finished. In
         *    this case, we ignore this transaction because it would
         *    not be visible to the current snapshot.
         */
        while (s != txn_shared && (id = s->id) != WT_TXN_NONE && WT_TXNID_LE(prev_oldest_id, id) &&
          WT_TXNID_LT(id, current_id)) {
            /*
             * If the transaction is still allocating its ID, then we spin here until it gets its
             * valid ID.
             */
            WT_READ_BARRIER();
            if (!s->is_allocating) {
                /*
                 * There is still a chance that fetched ID is not valid after ID allocation, so we
                 * check again here. The read of transaction ID should be carefully ordered: we want
                 * to re-read ID from transaction state after this transaction completes ID
                 * allocation.
                 */
                WT_READ_BARRIER();
                if (id == s->id) {
                    txn->snapshot[n++] = id;
                    if (WT_TXNID_LT(id, pinned_id))
                        pinned_id = id;
                    break;
                }
            }
            WT_PAUSE();
        }
    }
```
There are barriers in ____txn_get_snapshot_int, but I'm confused how that relates to this issue? I think setting a volatile bool value should always be atomic.

Comment (luke-pearson) [reply src/evict/evict_page.c:684]: As I understand it an operation using a volatile isn't guaranteed to be atomic, but the read itself is guaranteed to re-read the memory location of the data instead of using a cached version.

Comment (quchenhao) [reply src/evict/evict_page.c:684]: Yes, but assigning a value to a 64 bit memory location should be atomic on a 64 bit machine?

Comment (luke-pearson) [reply src/evict/evict_page.c:684]: I think in a multi-threaded context the answer is no? But I'm a bit unsure at this point.

Comment (hbokhari) [reply src/evict/evict_page.c:684]: I think you need to revisit the fact that barriers are not for ensuring atomicity - they are there to ensure read and writes are not shuffled around either by the compiler or the processor hardware (which is designed to execute instructions out of order). Read barriers ensures that all previous read instructions have executed before executing the next instruction.

Comment (hbokhari) [reply src/evict/evict_page.c:684]: `volatile` is used to indicate that variable value can change in some other thread so don't cache the copy in internal register. For example:
`for(;wait = true;)`
`   printf("Waiting ...\n"); 
`   sleep(10);`
In the example above, if we are declaring the `wait` variable with `volatile` keyword, the compiler will cache the local copy of `wait` and we will keep looping independently although some other thread might have set the `wait` to false.

Comment (hbokhari) [reply src/evict/evict_page.c:684]: I added the comment here indicating that we have taken the decision to use snapshot because the `checkpoint_running` was false at time. However, sometime between then and now, the variable `checkpoint_running` is set to true and we have to revert whatever preparation we have done.

Comment (quchenhao) [reply src/evict/evict_page.c:684]: Yes, I agree with Hasseb. Here we only have one state so we should not need any barrier. In addition, read write of any memory data under 64 bit should be atomic on a 64 bit machine.

Comment (luke-pearson) [reply src/evict/evict_page.c:684]: I still believe we need a barrier here however when I made the original comment I didn't check for one in the bump snapshot code, which should suffice.

Comment (hbokhari) [reply src/evict/evict_page.c:684]: My opinion here is that we need to have a barrier somewhere, which might as well be in bump snapshot code. But my only concern is that our decision of not using a barrier is based on a function code that may chance in future.

Comment (quchenhao) [reply src/evict/evict_page.c:684]: I think if we need a barrier, it should be a write barrier when we set the global checkpoint_running flag not when we read it.

Comment (agorrod) [reply src/evict/evict_page.c:684]: I'm not sure we have a shared understanding of barriers. Could we have a chat at 4?

File: src/evict/evict_page.c
Comment (agorrod) [reply src/evict/evict_page.c:698]: ack - my mistake. Thanks for explaining.

**Review by agorrod — COMMENTED** [review]
I'm good with this change now, modulo attempting to make the logic around setting the new local `checkpoint_running` variable in a safer way.

**Comment by agorrod** [comment]
Don't wait on me for this PR. I'm happy when the other reviewers are.

**Comment by quchenhao** [comment]
@luke-pearson @hbokhari @bvpvamsikrishna @agorrod This is ready for review again.

File: src/evict/evict_page.c
Comment (luke-pearson) [inline src/evict/evict_page.c:666]: Grammar: "History store data is always evictable."

File: src/evict/evict_page.c
Comment (luke-pearson) [inline src/evict/evict_page.c:668]: Comment formatting is wrong here also I think a TODO should be FIX-ME-TICKET_NUMBER format do we have a ticket to record here?

Comment (quchenhao) [reply src/evict/evict_page.c:668]: There should be one I think.

File: src/evict/evict_page.c
Comment (quchenhao) [reply src/evict/evict_page.c:666]: I thought the word data is plural. Found an interesting article discuss this https://ludwig.guru/blog/data-is-vs-data-are-an-intriguing-dilemma-with-a-surprising-twist/?gclid=CjwKCAjw_o-HBhAsEiwANqYhp_o85TtjLxJIIq11Ux4M1cvCw7lrhc7_rwGYn95QM-Zpy3N6swgkLRoCojEQAvD_BwE

Comment (luke-pearson) [reply src/evict/evict_page.c:666]: Huh fair enough, at first glance it does seem wrong to me so I guess the "data is" that i've come to expect is a modernization? I think we should go with the more modern form of "is" still but you've proven that what you wrote is correct anyway so your choice.

Comment (quchenhao) [reply src/evict/evict_page.c:666]: OK, I'll go for the popular one.

File: src/evict/evict_page.c
Comment (luke-pearson) [inline src/evict/evict_page.c:679]: I know this wasn't your change but can we fix this one too? "We have an exclusive lock", not "We have a exclusive lock"

Comment (quchenhao) [reply src/evict/evict_page.c]: done

File: src/evict/evict_page.c
Comment (luke-pearson) [inline src/evict/evict_page.c:666]: Can you add a warning comment here that changing these if cases could result in a unforeseen consequences or similar? We need to make it clear that this code you shouldn't change without a very well thought out reason.

Comment (luke-pearson) [reply src/evict/evict_page.c:666]: Something like "Take care when modifying the following logic as there are a number of edge cases here that aren't immediately obvious."

Comment (quchenhao) [reply src/evict/evict_page.c:666]: Do we have such kind of precedence? I think that just discourages people from making the changes.

Comment (luke-pearson) [reply src/evict/evict_page.c:666]: We do have similar comments in other places that are along the same lines, and yes the intention is to discourage people from making changes without a good idea of said edge case are.

Comment (quchenhao) [reply src/evict/evict_page.c:666]: The intention of the refactor is to make such cases obvious? Do you think if there is still any subtle edge cases following the new code flow?

Comment (luke-pearson) [reply src/evict/evict_page.c:666]: We are modifying this code to handle a race that can occur between eviction and checkpoint which results in the history store having records removed incorrectly. That case isn't mentioned in any of the following comments explicitly.

Comment (quchenhao) [reply src/evict/evict_page.c:666]: added

File: src/evict/evict_page.c
Comment (luke-pearson) [inline src/evict/evict_page.c:700]: I appreciate that you added this comment but can we not add the same comment in 4 places? Rather at the top of the the main if statement chain add a comment.

**Review by luke-pearson — APPROVED** [review]
LGTM.

File: src/evict/evict_page.c
Comment (hbokhari) [inline src/evict/evict_page.c:716]: @quchenhao - there is still a chance that checkpoint will start running between checking the condition and setting the flag `WT_REC_APP_EVICTION_SNAPSHOT`. I wonder if we are okay with that scenario, and if yes, we need to update the comment saying so.

**Review by hbokhari — COMMENTED** [review]
See me comment below:

File: src/evict/evict_page.c
Comment (quchenhao) [reply src/evict/evict_page.c:716]: I think that is OK, as checkpoint will have a newer snapshot than eviction. I'll add the comment.

**Review by hbokhari — APPROVED** [review]
LGTM!

**Review by bvpvamsikrishna — COMMENTED** [review]
lgtm

**Pull request merged**
